File: neetcode/problems/leetcode20/solution.js
const complements = {
	')': '(',
	'}': '{',
	']': '[',
};

const isValid = function (s) {
	if (s.length % 2 !== 0) return false;

	const stack = [];

	for (let i = 0; i < s.length; i++) {
		const complement = complements[s[i]];

		if (!complement) stack.push(s[i]);
		else if (complement === stack.at(-1)) stack.pop();
		else return false;
	}

	return !!!stack.length;
};

File: neetcode/problems/leetcode20/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False

        complements = {")": "(", "}": "{", "]": "["}
        stack = []

        for p in s:
            if p in complements:
                if stack and stack[-1] == complements[p]:
                    stack.pop()
                else:
                    return False

            else:
                stack.append(p)

        return False if stack else True
